Rewrite to be based on php-build

Rusydy opened the pull request:
TLDR; to solve https://github.com/asdf-community/asdf-php/issues/198, https://github.com/asdf-community/asdf-php/issues/72 and https://github.com/asdf-community/asdf-php/issues/194

This pull request introduces significant improvements to the PHP plugin's development workflow, documentation, and compatibility with modern dependencies. The most notable changes include a complete overhaul of the GitHub Actions CI workflow for better cross-platform and legacy PHP support, new documentation for manual local testing, compatibility patches for ICU 74 and libxml2 2.12, and a script for installing OpenSSL 1.1. Additionally, a new submodule for `php-build` is added to the repository.

**CI/CD and Development Workflow Improvements**
- Major rewrite of `.github/workflows/workflow.yml` to use a single, matrix-based job supporting both Ubuntu and macOS, legacy and modern PHP versions, with improved dependency management and diagnostics. Adds linting and formatting jobs with more robust dependency installation and caching.
- Adds `vendor/php-build` as a submodule, ensuring consistent build tooling. (`.gitmodules`)

**Documentation Enhancements**
- Expands `README.md` with clearer installation, update, and usage instructions for the asdf PHP plugin.
- Adds `docs/local-testing-manually.md` with detailed, step-by-step guides for installing, uninstalling, and troubleshooting various PHP versions locally, including custom build flags and log monitoring.

**Compatibility and Build Fixes**
- Adds patch for PHP 7.4 to fix compatibility with ICU 74 (`patches/php-7.4-icu-74-compat.patch`).
- Adds patch for PHP 8.0 to fix compatibility with ICU 74 (`patches/php-8.0-icu-74-compat.patch`).
- Adds patch for PHP 8.0 to fix compatibility with libxml2 2.12 (`patches/php-8.0-libxml2-2.12-compat.patch`).

**Utility Scripts**
- Introduces `lib/install-openssl11.sh`, a robust script for building and installing OpenSSL 1.1 locally, with dependency checks, colored output, and instructions for environment configuration.


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Modernizes the plugin around `php-build`, expanding CI/test coverage and improving legacy compatibility.
> 
> - Add `vendor/php-build` submodule and `bin/post-plugin-add` to auto-initialize it; new `bin/latest-stable` for resolving `latest`
> - Rewrite `bin/install` to drive `php-build`, auto-detect/auto-install OpenSSL 1.1 (`lib/install-openssl11.sh`), set compilers/flags for legacy PHP, install Composer, optional Xdebug, and patch definitions for ICU 74 and libxml2 2.12
> - Add runtime env tweaks in `bin/exec-env` to set `LD_LIBRARY_PATH` for OpenSSL 1.1 when required
> - Overhaul CI workflow to a single matrix job across Ubuntu/macOS and legacy/modern versions; add `lint` and `format` jobs
> - Add compatibility patches: `patches/php-7.4-icu-74-compat.patch`, `patches/php-8.0-icu-74-compat.patch`, `patches/php-8.0-libxml2-2.12-compat.patch`
> - Update `README.md` and add `docs/local-testing-manually.md`; add `.gitmodules` and expand `.gitignore`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 349250600b3f2994079ae8532561cb082b36544f. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

Rusydy commented:
Waiting for https://github.com/php-build/php-build/pull/796 to be approved by php-build's maintainer

Rusydy commented:
## Installation test of 7.4.14 on Almalinux 10
<img width="1512" height="773" alt="Screenshot 2025-12-17 at 22 26 39" src="https://github.com/user-attachments/assets/2f65b0da-10b5-4204-81ec-217e739c4d09" />

## Installation test of 8.0.0 on Almalinux 10
<img width="1506" height="800" alt="Screenshot 2025-12-17 at 21 26 22" src="https://github.com/user-attachments/assets/309300ec-3361-46a4-88c0-3a8a3187d2f6" />

## Installation test of 8.3.29 on Almalinux 10
<img width="1512" height="824" alt="Screenshot 2025-12-26 at 04 12 28" src="https://github.com/user-attachments/assets/dd58cb34-49e9-4bcb-af51-acef1b2d297c" />

Rusydy committed:
Refactor: Complete rewrite to use php-build instead of custom download/compile

- Replace custom download and compilation logic with php-build
- Add automatic OpenSSL 1.1 compatibility detection and installation
- Add compiler warning fixes for modern GCC versions
- Add automated Composer installation with signature verification
- Add Xdebug installation support
- Add comprehensive patches for PHP 7.4 and 8.0 compatibility
- Add post-plugin-add hook for automatic php-build initialization
- Add latest-stable version resolution
- Add extensive documentation and testing instructions
- Add support for all php-build environment variables and options

This addresses build failures on modern systems and provides a more
maintainable and feature-rich PHP installation experience.

Rusydy committed:
docs: Expand local testing instructions with latest, 8.5.1, and 8.3.29 examples

- Add comprehensive installation examples for latest stable version
- Include verbose installation options and custom compiler flags
- Add monitoring instructions with log file tailing
- Improve command consistency using 'asdf plugin add php $(pwd)'
- Document verification steps for successful installations

Rusydy committed:
fix: Apply shfmt formatting to shell scripts

- Remove spaces around redirection operators (&>/dev/null, >/dev/null, >>, etc.)
- Conform to shfmt standard formatting rules (-s -i 2 -ci)
- Resolves CI format check failure

Rusydy committed:
fix: Resolve shellcheck warnings for improved code quality

- Fix SC2155: Separate variable declaration and assignment to avoid masking return codes
- Fix SC2086: Add proper quoting for variables to prevent word splitting
- Fix SC2034: Remove unused php_status variable
- Apply fixes to bin/install and bin/exec-env
- All shellcheck warnings resolved while maintaining functionality

Rusydy committed:
Fix sed command and executable permissions for bin/install and bin/exec-env

Rusydy committed:
Fix Composer installation and make OpenSSL installation non-interactive in CI

Rusydy committed:
Update CI matrix with compatible PHP versions and build environment

Rusydy committed:
Add install monitoring and verification steps

Rusydy committed:
Replace plugin-test action with custom workflow to handle submodule and permissions

Rusydy committed:
Add timeout and better error handling to CI workflow

Rusydy committed:
Add verbose logging for troubleshooting CI failures

Rusydy committed:
Fix asdf setup and plugin installation conflicts in CI

Rusydy committed:
Consolidate CI steps into single script to avoid shell environment issues

Rusydy committed:
Fix shell environment and improve asdf setup in CI

Rusydy committed:
Rename workflow step and add setup logs

Rusydy committed:
Use asdf actions and simplify CI workflow

Replace manual asdf/PHP setup with asdf-vm/actions/plugin-test and
remove ad-hoc asdf install and plugin steps. Add a non-fatal Composer
version check and initialize git submodules earlier, setting executable
permissions. Use asdf-vm/actions/setup and asdf-vm/actions/install to
provide shfmt, shellcheck and make, and simplify lint/format commands.

Rusydy committed:
Replace asdf actions with direct tool installs

Use apt-get to install shellcheck and make and download shfmt v3.8.0
Drop asdf-vm actions from workflow

Rusydy committed:
Pass git URL and ref into PHP setup step

Provide giturl and gitref environment inputs so the PHP setup action can
access the current repository and checkout the exact commit.

Rusydy committed:
Replace plugin-test with manual asdf setup

Add asdf setup action and install the plugin from the current checkout,
then install and set the PHP version globally. Increase the test step
timeout to 30 minutes.

Rusydy committed:
Update workflow.yml

Rusydy committed:
Improve compiler setup for old PHP versions

Add setup_compiler_for_old_php to choose gcc-9 when available and export
CC/CXX and autoconf cache variables; set conservative
CFLAGS/CXXFLAGS/LDFLAGS for PHP 7.4 and PHP 8.0.<20. Add
validate_compiler_setup to verify the compiler and a
prepare_php_build_environment wrapper that preserves the environment
when invoking php-build. Use the wrapper and run compiler tests for
problematic versions, falling back to the normal build path otherwise

Rusydy committed:
Update workflow.yml

Rusydy committed:
Fix shell test grouping and clean up spacing

Use brace groups for compound [ ... ] tests to avoid relying on subshell
parentheses and improve POSIX shell compatibility. Remove unnecessary
spaces around redirections and here-docs and tidy a minor comment.

Rusydy committed:
Only set conservative build flags if unset

Use default-only parameter expansion so existing CFLAGS, CXXFLAGS,
LDFLAGS, PHP_BUILD_CONFIGURE_OPTS and MAKE_OPTS are not overridden or
appended to. Add --without-tidy to the default PHP_BUILD_CONFIGURE_OPTS
for older PHP versions.

Rusydy committed:
Update workflow.yml

Rusydy committed:
Remove trailing whitespace in bin/install

Rusydy committed:
Update workflow.yml

Rusydy committed:
Improve Composer installer robustness and logging

Rusydy committed:
Update workflow.yml

Rusydy committed:
Refactor CI workflow and unify setup

Add shared env lists for Ubuntu and macOS packages, switch the matrix to
include entries with an is_legacy flag, and unify system package
installation into a single step. Consolidate PHP configuration and
installation (exporting vars to GITHUB_ENV, legacy compiler handling and
diagnostics), add job timeout, and simplify lint/format steps with minor
improvements (curl -sL).

Rusydy committed:
Update workflow.yml

Rusydy commented:
Finish work on macOS latest with PHP 8.3.29 pipeline as of 2025-12-27 

<img width="853" height="678" alt="Screenshot 2025-12-27 at 23 43 17" src="https://github.com/user-attachments/assets/08cc3584-44aa-4f3f-96a8-4205dd54524b" />

continue for other version

Rusydy committed:
Improve macOS PHP build and fix mktemp usage

Add bzip2 to MACOS_DEPS and mark macOS 8.3.29 as legacy Set CXXFLAGS,
PKG_CONFIG_PATH and ICU prefix fallback for macOS Export
PHP_BUILD_CONFIGURE_OPTS to use bzip2, libiconv and openssl@3 Use
portable mktemp /tmp/test_XXXXXX.c instead of --suffix=.c

Rusydy committed:
Update workflow.yml

Rusydy commented:
## Prerequisite

```bash
# Clean up any previous failed attempts first
rm -rf /var/tmp/php-build/source/8.3.29
```

## Installation test of 8.3.29 on local macOS Tahoe 26.1

```bash
# Set the same environment variables that worked with php-build directly
export CXXFLAGS="-std=c++17 -stdlib=libc++"
export PKG_CONFIG_PATH="$(brew --prefix icu4c@78)/lib/pkgconfig:$(brew --prefix openssl@3)/lib/pkgconfig:$(brew --prefix libxml2)/lib/pkgconfig:$PKG_CONFIG_PATH"
export PHP_BUILD_CONFIGURE_OPTS="--with-bz2=$(brew --prefix bzip2) --with-iconv=$(brew --prefix libiconv) --with-openssl=$(brew --prefix openssl@3) --with-icu-dir=$(brew --prefix icu4c@78)"

# Now run asdf install
asdf install php 8.3.29
```

<img width="1512" height="908" alt="Screenshot 2025-12-27 at 23 31 24" src="https://github.com/user-attachments/assets/78649eea-ab96-4b61-ac73-2d461372d221" />

smorimoto commented:
bugbot run

Rusydy committed:
Make sed -i usage cross-platform

Rusydy committed:
Default php_patch to 0 for comparisons

Rusydy committed:
Remove unnecessary quotes in OSTYPE test

Rusydy commented:
bugbot run

smorimoto commented:
bugbot run

Rusydy committed:
Escape variables in generated wrapper script

Rusydy committed:
Use array for sed -i to ensure portability

Quote OSTYPE in the comparison and store sed command as an array. This
allows passing an empty -i argument on macOS (BSD sed) and running sed
consistently via "${SED_INPLACE[@]}" for both macOS and GNU sed.

Rusydy committed:
Remove unnecessary quotes from OSTYPE check

Rusydy committed:
Add SHA256 verification and --help to installer

Support an OPENSSL_SHA256 environment variable and attempt to retrieve
the official checksum from OpenSSL/GitHub when unset. Verify downloads
using sha256sum or shasum, exit on mismatch, and warn if no verification
tool or checksum is available. Add a --help flag with usage examples and
improve flag handling.

Rusydy committed:
Consolidate macOS brew setup in workflow

Set ICU_PREFIX (prefer icu4c@78) and export PKG_CONFIG_PATH. Detect
tidy-html5 once and export TIDY_CONFIG as --with-tidy or --without-tidy,
then reuse it in PHP_BUILD_CONFIGURE_OPTS. Remove debug tidy install
logging and duplicate configure blocks.

Rusydy committed:
Update README with architecture and setup

Rusydy committed:
Update workflow.yml

Rusydy commented:
I am currently focused on PHP 8.0.0 installation on macOS

Rusydy committed:
Fix BSD sed compatibility by replacing sed append commands with awk

- Replace all sed 'a' (append) commands with portable awk equivalents
- Fixes macOS BSD sed error: 'command a expects \ followed by text'
- Remove unused SED_INPLACE variable
- Maintain identical patch insertion functionality
- Cross-platform compatible (macOS, Linux, BSD)

Rusydy committed:
Fix macOS-specific configuration for legacy PHP versions

- Add OS detection for platform-specific build flags
- Configure BZip2 and libiconv paths for macOS using Homebrew
- Remove incompatible Linux linker flags (--no-as-needed) on macOS
- Set proper LDFLAGS and CPPFLAGS for Homebrew library paths
- Fixes BZip2 'Please reinstall the BZip2 distribution' error on macOS

Rusydy committed:
Create separate temp source and exe files

Rusydy committed:
Simplify shell conditionals and redirects

OliverWich commented:
Excited to finally have asdf (and thus mise) be able to install php properly again.

Thanks for your work, looking forward to when this reaches `master`, keep it up!

Rusydy commented:
I am still working on PHP 8.0.0 installation on macOS

below is my note for tracking down the issue using `plainuml`
```plainuml
@startuml
start

:Pemeriksaan Platform & Arsitektur;
:Inisialisasi Jalur php-build Internal;
:Muat Definisi Versi;
    if (Versi Tersedia?) is (<color:black>yes) then
    :Panggil php-build Executable;
    :\t\t\t\t\t\t; <<timeEvent>>
        if (Kompilasi Berhasil?) is (<color:black>yes) then
            :Mapping Binary ke Path ASDF;
        else (<color:red>no)
            :Tampilkan Log Error dari php-build;
            stop
        endif
        :Registrasi Shims;
        stop
    else (<color:red>no)
        :Tampilkan Pesan Error;
        stop
    endif
@enduml
```

2026 January 24; turns out the error is in php-build, I need to work on php-build first

OliverWich commented:
If I can provide any help, I am happy to help / test on Arch Linux.

Just let me know what to do :)

Rusydy commented:
Hi @OliverWich 

Yes, I would like to have your perspective about these changes on Arch Linux.
You can see the step by step to test it on `docs/local-testing-manually.md`
Currently, I only test them on macOS, Alma Linux and Ubuntu

Rusydy commented:
One of my PR, https://github.com/php-build/php-build/pull/796#event-22648343328, has been merged
I will continue to investigate

and I will add CI for PHP 8.5.2 which released in January 2026

Rusydy-Mekari committed:
Update local testing docs for PHP plugin install

Update example PHP version comment to 8.5.4 (2026-04-07) and replace the
symlink-based plugin setup with 'asdf plugin add php $(pwd)' to use the
current directory

Rusydy-Mekari committed:
Add macOS-specific dependency configuration for PHP builds

Rusydy-Mekari committed:
Add macOS-specific tests and verification for bzip2 configuration

Rusydy-Mekari committed:
Handle Homebrew keg-only prefixes and CXXFLAGS

Rusydy-Mekari committed:
Extract correct function in test setup

Rusydy-Mekari committed:
Require C++17 for intl extension

Update ext/intl/config.m4 to compile with C++17.
ICU 74+ headers use C++17 features (enable_if_t, void_t, is_same_v,
if constexpr), so PHP_CXX_COMPILE_STDCXX was bumped from 11 to 17

Rusydy-Mekari committed:
Fix C++17/ICU build issues for old PHP

Add -Wno-int-conversion to CFLAGS for PHP 7.4 and 8.0.x to silence int
conversion warnings.
Append to PHP_BUILD_CONFIGURE_OPTS and LDFLAGS instead of only setting
defaults so existing values are preserved.
On macOS enable CXXFLAGS -std=c++17 -stdlib=libc++
-DU_USING_ICU_NAMESPACE and prefer Homebrew's libgd via
--with-external-gd when available.
Inject an intl C++17 patch into php-build definitions to require C++17
for ext/intl so ICU 74+/78+ headers compile correctly.

Rusydy-Mekari committed:
Use ASDF_CONCURRENCY=8 for PHP installs

Rusydy-Mekari committed:
Improve macOS diagnostics and idempotent flags

Rusydy committed:
[TICKET-01]: delete redundant validate_compiler_setup

Rusydy committed:
[TICKET-01]: delete prepare_php_build_environment wrapper

Rusydy committed:
[TICKET-01]: delete verify_bzip2_fix.sh and verify target

Rusydy committed:
[TICKET-05.1]: Update MACOS_DEPS to use OpenSSL 3 only

Rusydy committed:
[TICKET-05.2]: Fix type casting for xmlStructuredErrorFunc in libxml.c

Rusydy committed:
[TICKET-05.2]: inject php-8.0-libxml2-const-handler.patch for PHP 8.0.x builds

Rusydy committed:
[TICKET-05.3]: inject php-8.0-libxml2-attribute-unused.patch for PHP 8.0.x builds

Rusydy committed:
[TICKET-05.4]: inject php-8.0-bcmath-knr-to-ansi.patch for PHP 8.0.x
builds

Apple Clang 17 (Xcode 26) removed K&R C support entirely. PHP 8.0.0's
bcmath/libbcmath has 14 files using K&R-style function declarations
(parameter names in signature, types in separate declarations before {).

Converts all affected functions to ANSI C style (types inside parameter
list), which is the only syntax Clang 17 accepts.

Rusydy committed:
[TICKET-05.5]: Add tidy-html5 support in macOS dependencies configuration

Rusydy committed:
[TICKET-05.5]: Refactor GD image function pointer types for PHP 8.0 compatibility

Rusydy committed:
[TICKET-05.5]: Add GD function pointer types patch for PHP 8.0 compatibility

Rusydy committed:
[TICKET-05.7]: Add intl K&R to ANSI C patch for PHP 8.0 compatibility

Rusydy committed:
[TICKET-05.9]: Add pcntl function pointer type patch for PHP 8.0 compatibility
